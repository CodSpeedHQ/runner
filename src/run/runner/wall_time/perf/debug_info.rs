use crate::prelude::*;
use crate::run::runner::wall_time::perf::perf_map::ModuleSymbols;
use libc::pid_t;
use object::{Object, ObjectSection};
use serde::{Deserialize, Serialize};
use std::{fmt::Debug, path::Path};

#[derive(Hash, PartialEq, Eq, Clone, Serialize, Deserialize)]
pub struct DebugInfo {
    pub addr: u64,
    pub size: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub file: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub line: Option<u32>,
}

impl Debug for DebugInfo {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let location = match (&self.file, self.line) {
            (Some(file), Some(line)) => format!("{file}:{line}"),
            (Some(file), None) => file.clone(),
            _ => String::from("<unknown>"),
        };
        let name = self.name.as_deref().unwrap_or("<no symbol>");
        write!(
            f,
            "DebugInfo {{ addr: {:x}, size: {:x}, name: {}, location: {} }}",
            self.addr, self.size, name, location
        )
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModuleDebugInfo {
    load_bias: u64,
    debug_infos: Vec<DebugInfo>,
}

impl ModuleDebugInfo {
    /// Create debug info from existing symbols by looking up file/line in DWARF
    pub fn from_symbols<P: AsRef<Path>>(path: P, symbols: &ModuleSymbols) -> anyhow::Result<Self> {
        let content = std::fs::read(path.as_ref())?;
        let object = object::File::parse(&*content)?;

        let ctx = Self::create_dwarf_context(&object).context("Failed to create DWARF context")?;
        let load_bias = symbols.load_bias();
        let debug_infos = symbols
            .symbols()
            .iter()
            .filter_map(|symbol| {
                let (file, line) = match ctx.find_location(symbol.addr) {
                    Ok(Some(location)) => {
                        let file = location.file.map(|f| f.to_string())?;
                        (file, location.line)
                    }
                    _ => return None,
                };

                Some(DebugInfo {
                    addr: symbol.addr,
                    size: symbol.size,
                    name: Some(symbol.name.clone()),
                    file: Some(file),
                    line,
                })
            })
            .collect();

        Ok(Self {
            load_bias,
            debug_infos,
        })
    }

    fn create_dwarf_context(
        object: &object::File,
    ) -> anyhow::Result<addr2line::Context<gimli::EndianRcSlice<gimli::RunTimeEndian>>> {
        let endian = if object.is_little_endian() {
            gimli::RunTimeEndian::Little
        } else {
            gimli::RunTimeEndian::Big
        };

        type EndianRcSlice = gimli::EndianRcSlice<gimli::RunTimeEndian>;
        let load_section = |id: gimli::SectionId| -> Result<EndianRcSlice, gimli::Error> {
            let data = object
                .section_by_name(id.name())
                .and_then(|s| s.uncompressed_data().ok())
                .unwrap_or(std::borrow::Cow::Borrowed(&[]));
            Ok(gimli::EndianRcSlice::new(
                std::rc::Rc::from(data.as_ref()),
                endian,
            ))
        };

        let dwarf = gimli::Dwarf::load(load_section)?;
        addr2line::Context::from_dwarf(dwarf).map_err(Into::into)
    }
}

/// Represents all the modules inside a process and their debug info.
pub struct ProcessDebugInfo {
    pid: pid_t,
    modules: Vec<ModuleDebugInfo>,
}

impl ProcessDebugInfo {
    pub fn new(
        pid: pid_t,
        process_symbols: &crate::run::runner::wall_time::perf::perf_map::ProcessSymbols,
    ) -> Self {
        let mut modules = Vec::new();
        for (path, module_symbols) in process_symbols.modules_with_symbols() {
            match ModuleDebugInfo::from_symbols(path, module_symbols) {
                Ok(module_debug_info) => {
                    modules.push(module_debug_info);
                }
                Err(error) => {
                    debug!("Failed to load debug info for module {path:?}: {error}");
                }
            }
        }

        Self { pid, modules }
    }

    pub fn save_to<P: AsRef<std::path::Path>>(&self, folder: P) -> anyhow::Result<()> {
        if self.modules.is_empty() {
            return Ok(());
        }

        let debug_info_path = folder
            .as_ref()
            .join(format!("perf-{}.debuginfo.json", self.pid));
        let json = serde_json::to_string(&self.modules)?;
        std::fs::write(debug_info_path, json)?;

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::run::runner::wall_time::perf::perf_map::ModuleSymbols;

    #[test]
    fn test_golang_debug_info() {
        let (start_addr, end_addr, file_offset) =
            (0x0000000000402000_u64, 0x000000000050f000_u64, 0x2000);
        let module_symbols = ModuleSymbols::new(
            "testdata/perf_map/go_fib.bin",
            start_addr,
            end_addr,
            file_offset,
        )
        .unwrap();
        let module_debug_info =
            ModuleDebugInfo::from_symbols("testdata/perf_map/go_fib.bin", &module_symbols).unwrap();
        insta::assert_debug_snapshot!(module_debug_info.debug_infos);
    }

    #[test]
    fn test_cpp_debug_info() {
        let (start_addr, end_addr, file_offset) =
            (0x0000000000400000_u64, 0x0000000000459000_u64, 0x0);
        let module_symbols = ModuleSymbols::new(
            "testdata/perf_map/cpp_my_benchmark.bin",
            start_addr,
            end_addr,
            file_offset,
        )
        .unwrap();
        let mut module_debug_info = ModuleDebugInfo::from_symbols(
            "testdata/perf_map/cpp_my_benchmark.bin",
            &module_symbols,
        )
        .unwrap();

        module_debug_info.debug_infos.sort_by_key(|d| d.addr);

        insta::assert_debug_snapshot!(module_debug_info.debug_infos);
    }

    #[test]
    fn test_rust_divan_debug_info() {
        const MODULE_PATH: &str = "testdata/perf_map/divan_sleep_benches.bin";

        let module_symbols =
            ModuleSymbols::new(MODULE_PATH, 0x00005555555a2000, 0x0000555555692000, 0x4d000)
                .unwrap();
        let module_debug_info =
            ModuleDebugInfo::from_symbols(MODULE_PATH, &module_symbols).unwrap();
        insta::assert_debug_snapshot!(module_debug_info.debug_infos);
    }

    #[test]
    fn test_the_algorithms_debug_info() {
        const MODULE_PATH: &str = "testdata/perf_map/the_algorithms.bin";

        let module_symbols = ModuleSymbols::new(
            MODULE_PATH,
            0x00005573e59fe000,
            0x00005573e5b07000,
            0x00052000,
        )
        .unwrap();
        let module_debug_info =
            ModuleDebugInfo::from_symbols(MODULE_PATH, &module_symbols).unwrap();
        insta::assert_debug_snapshot!(module_debug_info.debug_infos);
    }

    #[test]
    fn test_ruff_debug_info() {
        const MODULE_PATH: &str = "testdata/perf_map/ty_walltime";

        let (start_addr, end_addr, file_offset) =
            (0x0000555555e6d000_u64, 0x0000555556813000_u64, 0x918000);
        let module_symbols =
            ModuleSymbols::new(MODULE_PATH, start_addr, end_addr, file_offset).unwrap();
        let module_debug_info =
            ModuleDebugInfo::from_symbols(MODULE_PATH, &module_symbols).unwrap();
        insta::assert_debug_snapshot!(module_debug_info.debug_infos);
    }
}
