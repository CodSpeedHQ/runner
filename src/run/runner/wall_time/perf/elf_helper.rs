//! Based on this: https://github.com/mstange/samply/blob/4a5afec57b7c68b37ecde12b5a258de523e89463/samply/src/linux_shared/svma_file_range.rs#L8

use anyhow::Context;
use object::Object;
use object::ObjectSegment;

// A file range in an object file, such as a segment or a section,
// for which we know the corresponding Stated Virtual Memory Address (SVMA).
struct SvmaFileRange {
    pub svma: u64,
    pub file_offset: u64,
    pub size: u64,
}

impl SvmaFileRange {
    pub fn from_segment<'data, S: ObjectSegment<'data>>(segment: S) -> Self {
        let svma = segment.address();
        let (file_offset, size) = segment.file_range();
        SvmaFileRange {
            svma,
            file_offset,
            size,
        }
    }

    pub fn encompasses_file_range(&self, runtime_file_offset: u64, mapping_size: u64) -> bool {
        self.file_offset <= runtime_file_offset
            && (runtime_file_offset + mapping_size) <= (self.file_offset + self.size)
    }

    pub fn is_encompassed_by_file_range(
        &self,
        runtime_file_offset: u64,
        mapping_size: u64,
    ) -> bool {
        runtime_file_offset <= self.file_offset
            && (self.file_offset + self.size) <= (runtime_file_offset + mapping_size)
    }
}

pub fn compute_load_bias(
    runtime_start_addr: u64,
    runtime_end_addr: u64,
    runtime_file_offset: u64,
    object: &object::File,
) -> anyhow::Result<u64> {
    // The addresses of symbols read from an ELF file on disk are not their final runtime addresses.
    // This is due to Address Space Layout Randomization (ASLR) and the way the OS loader maps
    // file segments into virtual memory.
    //
    // Step 1: Find the corresponding ELF segment.
    // We must find the `PT_LOAD` segment that corresponds to the executable memory region we found
    // in /proc/<pid>/maps. We do this by comparing the `runtime_offset` against the offset in the file.
    //
    // For example, if we have the following `/proc/<pid>/maps` output:
    // ```
    // 00400000-00402000 r--p 00000000 fe:01 114429641            /runner/testdata/perf_map/go_fib.bin
    // 00402000-0050f000 r-xp 00002000 fe:01 114429641            /runner/testdata/perf_map/go_fib.bin      <-- we find this
    // 0050f000-0064b000 r--p 0010f000 fe:01 114429641            /runner/testdata/perf_map/go_fib.bin
    // 0064b000-0064c000 r--p 0024a000 fe:01 114429641            /runner/testdata/perf_map/go_fib.bin
    // 0064c000-0065e000 rw-p 0024b000 fe:01 114429641            /runner/testdata/perf_map/go_fib.bin
    // 0065e000-00684000 rw-p 00000000 00:00 0
    // ```
    //
    // We'll match the PT_LOAD segment with the same offset (0x2000):
    // ```
    // $ readelf -l testdata/perf_map/go_fib.bin
    // Elf file type is EXEC (Executable file)
    // Entry point 0x402490
    // There are 15 program headers, starting at offset 64
    //
    // Program Headers:
    //   Type           Offset             VirtAddr           PhysAddr
    //   PHDR           0x0000000000000040 0x0000000000400040 0x0000000000400040
    //                  0x0000000000000348 0x0000000000000348  R      0x8
    //   INTERP         0x0000000000000430 0x0000000000400430 0x0000000000400430
    //                  0x0000000000000053 0x0000000000000053  R      0x1
    //   LOAD           0x0000000000000000 0x0000000000400000 0x0000000000400000
    //                  0x0000000000001640 0x0000000000001640  R      0x1000
    //   LOAD           0x0000000000002000 0x0000000000402000 0x0000000000402000        <-- we'll match this
    //                  0x000000000010ceb1 0x000000000010ceb1  R E    0x1000
    // ```
    let mapping_size = runtime_end_addr - runtime_start_addr;
    let load_segment = object
        .segments()
        .map(SvmaFileRange::from_segment)
        .find(|segment| {
            // When the kernel loads an ELF file, it maps entire pages (usually 4KB aligned),
            // not just the exact segment boundaries. Here's what happens:
            //
            // **ELF File Structure**:
            // - LOAD segment 1: file offset 0x0      - 0x4d26a  (data/code)
            // - LOAD segment 2: file offset 0x4d26c  - 0x13c4b6 (executable code)
            //
            // **Kernel Memory Mapping**: The kernel rounds down to page boundaries when mapping:
            // - Maps pages starting at offset 0x0     (covers segment 1)
            // - Maps pages starting at offset 0x4d000 (page-aligned, covers segment 2)
            //
            // (the example values are based on the `test_rust_divan_symbols` test)
            segment.encompasses_file_range(runtime_file_offset, mapping_size)
                || segment.is_encompassed_by_file_range(runtime_file_offset, mapping_size)
        })
        .context(format!(
            "Could not find segment or section overlapping the file offset range 0x{:x}..0x{:x}",
            runtime_file_offset,
            runtime_file_offset + mapping_size
        ))?;

    // Compute the actual virtual address at which the segment is located in process memory.
    let runtime_start_addr = if load_segment.file_offset > runtime_file_offset {
        runtime_start_addr + (load_segment.file_offset - runtime_file_offset)
    } else {
        runtime_start_addr - (runtime_file_offset - load_segment.file_offset)
    };

    // Step 2: Calculate the "load bias".
    // The bias is the difference between where the segment *actually* is in memory versus where the
    // ELF file *preferred* it to be.
    //
    //   load_bias = runtime_start_addr - segment_preferred_vaddr
    //
    //  - `runtime_start_addr`: The actual base address of this segment in memory (from `/proc/maps`).
    //  - `load_segment.address()`: The preferred virtual address (`p_vaddr`) from the ELF file itself.
    //
    // This single calculation correctly handles both PIE/shared-objects and non-PIE executables:
    //  - For PIE/.so files:   `0x7f... (random) - 0x... (small) = <large_bias>`
    //  - For non-PIE files: `0x402000 (fixed) - 0x402000 (fixed) = 0`
    Ok(runtime_start_addr.wrapping_sub(load_segment.svma))
}

/// The "relative address base" is the base address which [`LookupAddress::Relative`]
/// addresses are relative to. You start with an SVMA (a stated virtual memory address),
/// you subtract the relative address base, and out comes a relative address.
///
/// This function computes that base address. It is defined as follows:
///
///  - For Windows binaries, the base address is the "image base address".
///  - For mach-O binaries, the base address is the vmaddr of the __TEXT segment.
///  - For ELF binaries, the base address is the vmaddr of the *first* segment,
///    i.e. the vmaddr of the first "LOAD" ELF command.
///
/// In many cases, this base address is simply zero:
///
///  - ELF images of dynamic libraries (i.e. not executables) usually have a
///    base address of zero.
///  - Stand-alone mach-O dylibs usually have a base address of zero because their
///    __TEXT segment is at address zero.
///  - In PDBs, "RVAs" are relative addresses which are already relative to the
///    image base.
///
/// However, in the following cases, the base address is usually non-zero:
///
///  - The "image base address" of Windows binaries is usually non-zero.
///  - mach-O executable files (not dylibs) usually have their __TEXT segment at
///    address 0x100000000.
///  - mach-O libraries in the dyld shared cache have a __TEXT segment at some
///    non-zero address in the cache.
///  - ELF executables can have non-zero base addresses, e.g. 0x200000 or 0x400000.
///  - Kernel ELF binaries ("vmlinux") have a large base address such as
///    0xffffffff81000000. Moreover, the base address seems to coincide with the
///    vmaddr of the .text section, which is readily-available in perf.data files
///    (in a synthetic mapping called "[kernel.kallsyms]_text").
///
/// Credits: https://github.com/mstange/samply/blob/4a5afec57b7c68b37ecde12b5a258de523e89463/samply-symbols/src/shared.rs#L513-L566
pub fn relative_address_base(object_file: &object::File) -> u64 {
    use object::read::ObjectSegment;
    if let Some(text_segment) = object_file
        .segments()
        .find(|s| s.name() == Ok(Some("__TEXT")))
    {
        // This is a mach-O image. "Relative addresses" are relative to the
        // vmaddr of the __TEXT segment.
        return text_segment.address();
    }

    if let object::FileFlags::Elf { .. } = object_file.flags() {
        // This is an ELF image. "Relative addresses" are relative to the
        // vmaddr of the first segment (the first LOAD command).
        if let Some(first_segment) = object_file.segments().next() {
            return first_segment.address();
        }
    }

    // For PE binaries, relative_address_base() returns the image base address.
    object_file.relative_address_base()
}

pub fn compute_base_avma(
    runtime_start_addr: u64,
    runtime_end_addr: u64,
    runtime_file_offset: u64,
    object: &object::File,
) -> anyhow::Result<u64> {
    let bias = compute_load_bias(
        runtime_start_addr,
        runtime_end_addr,
        runtime_file_offset,
        object,
    )?;
    let base_svma = relative_address_base(object);
    Ok(base_svma.wrapping_add(bias))
}
