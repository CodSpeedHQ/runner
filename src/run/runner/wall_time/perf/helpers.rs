use crate::prelude::*;
use linux_perf_data::{PerfFileReader, PerfFileRecord, linux_perf_event_reader::EventRecord};
use std::collections::HashMap;

/// Tries to find the pid of the sampled process within a perf.data file.
pub fn find_pid<P: AsRef<std::path::Path>>(perf_file: P) -> anyhow::Result<i32> {
    let content = std::fs::read(perf_file.as_ref())?;
    let reader = std::io::Cursor::new(content);

    let PerfFileReader {
        mut record_iter,
        mut perf_file,
    } = PerfFileReader::parse_file(reader)?;

    let mut pid_freq = HashMap::new();

    // Only consider the first N events to reduce the performance impact. Certain benchmark libraries can generate
    // more than 100k for each benchmark, which can slow down the runner a lot. The highest chance of finding
    // different pids is in the first few events, where there's a possible overlap.
    const COUNT_FIRST_N: usize = 1000;
    let mut i = 0;

    while let Some(record) = record_iter.next_record(&mut perf_file)? {
        let PerfFileRecord::EventRecord { record, .. } = record else {
            continue;
        };

        let Ok(parsed_record) = record.parse() else {
            continue;
        };

        let EventRecord::Sample(event) = parsed_record else {
            continue;
        };

        // Ignore kernel events
        if event.pid == Some(-1) {
            continue;
        }

        if let Some(pid) = event.pid {
            *pid_freq.entry(pid).or_insert(0) += 1;

            i += 1;
            if i >= COUNT_FIRST_N {
                break;
            }
        }
    }
    debug!("Pid frequency: {pid_freq:?}");

    // Choose the pid with the highest frequency. However, we can only use a pid if more than N% of the
    // events are from that pid.
    let total_count = pid_freq.values().sum::<u64>();
    let (pid, pid_count) = pid_freq
        .iter()
        .max_by_key(|&(_, count)| count)
        .ok_or_else(|| anyhow!("Couldn't find pid in perf.data"))?;
    log::debug!("Pid frequency: {pid_freq:?}");

    let pid_percentage = (*pid_count as f64 / total_count as f64) * 100.0;
    if pid_percentage < 75.0 {
        bail!(
            "Most common pid {} only has {:.2}% of total events",
            pid,
            pid_percentage
        );
    }

    Ok(*pid)
}
