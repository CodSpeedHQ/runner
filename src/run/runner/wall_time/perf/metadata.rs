// !!!!!!!!!!!!!!!!!!!!!!!!
// !! DO NOT TOUCH BELOW !!
// !!!!!!!!!!!!!!!!!!!!!!!!
// Has to be in sync with `perf-parser`.
//

use std::{collections::HashMap, path::Path};

use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct PerfMetadata {
    /// Name and version of the integration
    pub integration: (String, String),

    /// The URIs of the benchmarks in the order they were executed.
    pub bench_order_by_pid: HashMap<u32, Vec<String>>,
}

impl PerfMetadata {
    pub fn save_to<P: AsRef<Path>>(&self, path: P) -> anyhow::Result<()> {
        let file = std::fs::File::create(path.as_ref().join("perf.metadata"))?;
        serde_json::to_writer(file, self)?;
        Ok(())
    }
}
