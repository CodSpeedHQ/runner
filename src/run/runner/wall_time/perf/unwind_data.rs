//! WARNING: This file has to be in sync with perf-parser!

use crate::run::runner::wall_time::perf::elf_helper;
use anyhow::{Context, bail};
use debugid::CodeId;
use libc::pid_t;
use object::Object;
use object::ObjectSection;
use runner_shared::unwind_data::UnwindData;
use std::ops::Range;

pub trait UnwindDataExt {
    fn new(
        path_slice: &[u8],
        mapping_start_file_offset: u64,
        mapping_start_avma: u64,
        mapping_size: u64,
        build_id: Option<&[u8]>,
    ) -> anyhow::Result<Self>
    where
        Self: Sized;

    fn save_to<P: AsRef<std::path::Path>>(&self, folder: P, pid: pid_t) -> anyhow::Result<()>;
    fn to_file<P: AsRef<std::path::Path>>(&self, path: P) -> anyhow::Result<()>;
}

impl UnwindDataExt for UnwindData {
    // Based on this: https://github.com/mstange/linux-perf-stuff/blob/22ca6531b90c10dd2a4519351c843b8d7958a451/src/main.rs#L747-L893
    fn new(
        path_slice: &[u8],
        runtime_file_offset: u64,
        runtime_start_addr: u64,
        runtime_end_addr: u64,
        build_id: Option<&[u8]>,
    ) -> anyhow::Result<Self> {
        let avma_range = runtime_start_addr..runtime_end_addr;

        let path = String::from_utf8_lossy(path_slice).to_string();
        let Some(file) = std::fs::File::open(&path).ok() else {
            bail!("Could not open file {path}");
        };

        let mmap = unsafe { memmap2::MmapOptions::new().map(&file)? };
        let file = object::File::parse(&mmap[..])?;

        // Verify the build id (if we have one)
        match (build_id, file.build_id()) {
            (Some(build_id), Ok(Some(file_build_id))) => {
                if build_id != file_build_id {
                    let file_build_id = CodeId::from_binary(file_build_id);
                    let expected_build_id = CodeId::from_binary(build_id);
                    bail!(
                        "File {path:?} has non-matching build ID {file_build_id} (expected {expected_build_id})"
                    );
                }
            }
            (Some(_), Err(_)) | (Some(_), Ok(None)) => {
                bail!("File {path:?} does not contain a build ID, but we expected it to have one");
            }
            _ => {
                // No build id to check
            }
        };

        let base_avma = elf_helper::compute_base_avma(
            runtime_start_addr,
            runtime_end_addr,
            runtime_file_offset,
            &file,
        )?;
        let eh_frame = file.section_by_name(".eh_frame");
        let eh_frame_hdr = file.section_by_name(".eh_frame_hdr");

        fn section_data<'a>(section: &impl ObjectSection<'a>) -> Option<Vec<u8>> {
            section.data().ok().map(|data| data.to_owned())
        }

        let eh_frame_data = eh_frame.as_ref().and_then(section_data);
        let eh_frame_hdr_data = eh_frame_hdr.as_ref().and_then(section_data);

        fn svma_range<'a>(section: &impl ObjectSection<'a>) -> Range<u64> {
            section.address()..section.address() + section.size()
        }

        Ok(Self {
            path,
            avma_range,
            base_avma,
            eh_frame_hdr: eh_frame_hdr_data.context("Failed to find eh_frame hdr data")?,
            eh_frame_hdr_svma: eh_frame_hdr
                .as_ref()
                .map(svma_range)
                .context("Failed to find eh_frame hdr section")?,
            eh_frame: eh_frame_data.context("Failed to find eh_frame data")?,
            eh_frame_svma: eh_frame
                .as_ref()
                .map(svma_range)
                .context("Failed to find eh_frame section")?,
        })
    }

    fn save_to<P: AsRef<std::path::Path>>(&self, folder: P, pid: pid_t) -> anyhow::Result<()> {
        let unwind_data_path = folder.as_ref().join(format!(
            "{}_{:x}_{:x}.unwind",
            pid, self.avma_range.start, self.avma_range.end
        ));
        self.to_file(unwind_data_path)?;

        Ok(())
    }

    fn to_file<P: AsRef<std::path::Path>>(&self, path: P) -> anyhow::Result<()> {
        if let Ok(true) = std::fs::exists(path.as_ref()) {
            log::warn!(
                "{} already exists, file will be truncated",
                path.as_ref().display()
            );
            log::warn!("{} {:x?}", self.path, self.avma_range);
        }

        let mut writer = std::fs::File::create(path.as_ref())?;
        bincode::serialize_into(&mut writer, self)?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! assert_elf_load_bias {
        ($start_addr:expr, $end_addr:expr, $file_offset:expr, $module_path:expr, $expected_load_bias:expr) => {
            let expected_load_bias = $expected_load_bias as u64;

            let file_data = std::fs::read($module_path).expect("Failed to read test binary");
            let object = object::File::parse(&file_data[..]).expect("Failed to parse test binary");
            let load_bias =
                elf_helper::compute_load_bias($start_addr, $end_addr, $file_offset, &object)
                    .unwrap();
            println!("Load bias for {}: 0x{:x}", $module_path, load_bias);
            assert_eq!(
                load_bias, expected_load_bias,
                "Invalid load bias: {:x} != {:x}",
                load_bias, expected_load_bias
            );
        };
    }

    // Note: You can double-check the values by getting the /proc/<pid>/maps via gdb:
    // ```
    // $ gdb testdata/perf_map/<sample>.bin -ex "break main" -ex "run" -ex "info proc mappings" -ex "continue" -ex "quit" -batch
    // Start Addr         End Addr           Size               Offset             Perms File
    // 0x0000555555554000 0x00005555555a2000 0x4e000            0x0                r--p  /runner/testdata/perf_map/divan_sleep_benches.bin
    // 0x00005555555a2000 0x0000555555692000 0xf0000            0x4d000            r-xp  /runner/testdata/perf_map/divan_sleep_benches.bin
    // 0x0000555555692000 0x000055555569d000 0xb000             0x13c000           r--p  /runner/testdata/perf_map/divan_sleep_benches.bin
    // 0x000055555569d000 0x000055555569f000 0x2000             0x146000           rw-p  /runner/testdata/perf_map/divan_sleep_benches.bin
    // 0x00007ffff7c00000 0x00007ffff7c28000 0x28000            0x0                r--p  /nix/store/g8zyryr9cr6540xsyg4avqkwgxpnwj2a-glibc-2.40-66/lib/libc.so.6
    // 0x00007ffff7c28000 0x00007ffff7d9e000 0x176000           0x28000            r-xp  /nix/store/g8zyryr9cr6540xsyg4avqkwgxpnwj2a-glibc-2.40-66/lib/libc.so.6
    // 0x00007ffff7d9e000 0x00007ffff7df4000 0x56000            0x19e000           r--p  /nix/store/g8zyryr9cr6540xsyg4avqkwgxpnwj2a-glibc-2.40-66/lib/libc.so.6
    // 0x00007ffff7df4000 0x00007ffff7df8000 0x4000             0x1f3000           r--p  /nix/store/g8zyryr9cr6540xsyg4avqkwgxpnwj2a-glibc-2.40-66/lib/libc.so.6
    // 0x00007ffff7df8000 0x00007ffff7dfa000 0x2000             0x1f7000           rw-p  /nix/store/g8zyryr9cr6540xsyg4avqkwgxpnwj2a-glibc-2.40-66/lib/libc.so.6
    // 0x00007ffff7dfa000 0x00007ffff7e07000 0xd000             0x0                rw-p
    // 0x00007ffff7f8a000 0x00007ffff7f8d000 0x3000             0x0                rw-p
    // ...
    // ```

    #[test]
    fn test_golang_unwind_data() {
        const MODULE_PATH: &str = "testdata/perf_map/go_fib.bin";

        let (start_addr, end_addr, file_offset) =
            (0x0000000000402000_u64, 0x000000000050f000_u64, 0x2000);
        assert_elf_load_bias!(start_addr, end_addr, file_offset, MODULE_PATH, 0x0);
        insta::assert_debug_snapshot!(UnwindData::new(
            MODULE_PATH.as_bytes(),
            file_offset,
            start_addr,
            end_addr,
            None
        ));
    }

    #[test]
    fn test_cpp_unwind_data() {
        // gdb testdata/perf_map/cpp_my_benchmark.bin -ex "break main" -ex "run" -ex "info proc mappings" -ex "continue" -ex "quit" -batch
        // Start Addr         End Addr           Size               Offset             Perms File
        // 0x0000000000400000 0x0000000000459000 0x59000            0x0                r-xp  /home/not-matthias/Documents/work/wgit/runner/testdata/perf_map/cpp_my_benchmark.bin
        // 0x000000000045a000 0x000000000045b000 0x1000             0x59000            r--p  /home/not-matthias/Documents/work/wgit/runner/testdata/perf_map/cpp_my_benchmark.bin
        // 0x000000000045b000 0x000000000045c000 0x1000             0x5a000            rw-p  /home/not-matthias/Documents/work/wgit/runner/testdata/perf_map/cpp_my_benchmark.bin
        const MODULE_PATH: &str = "testdata/perf_map/cpp_my_benchmark.bin";

        let (start_addr, end_addr, file_offset) =
            (0x0000000000400000_u64, 0x0000000000459000_u64, 0x0);
        assert_elf_load_bias!(start_addr, end_addr, file_offset, MODULE_PATH, 0x0);

        insta::assert_debug_snapshot!(UnwindData::new(
            MODULE_PATH.as_bytes(),
            file_offset,
            start_addr,
            end_addr,
            None,
        ));
    }

    #[test]
    fn test_rust_divan_unwind_data() {
        const MODULE_PATH: &str = "testdata/perf_map/divan_sleep_benches.bin";

        let (start_addr, end_addr, file_offset) =
            (0x00005555555a2000_u64, 0x0000555555692000_u64, 0x4d000);
        assert_elf_load_bias!(
            start_addr,
            end_addr,
            file_offset,
            MODULE_PATH,
            0x555555554000
        );
        insta::assert_debug_snapshot!(UnwindData::new(
            MODULE_PATH.as_bytes(),
            file_offset,
            start_addr,
            end_addr,
            None
        ));
    }

    #[test]
    fn test_the_algorithms_unwind_data() {
        // $ gdb testdata/perf_map/the_algorithms.bin -ex "break main" -ex "run" -ex "info proc mappings" -ex "continue" -ex "quit" -batch
        // Start Addr         End Addr           Size               Offset             Perms File
        // 0x0000555555554000 0x00005555555a7000 0x53000            0x0                r--p  /home/not-matthias/Documents/work/wgit/runner/testdata/perf_map/the_algorithms.bin
        // 0x00005555555a7000 0x00005555556b0000 0x109000           0x52000            r-xp  /home/not-matthias/Documents/work/wgit/runner/testdata/perf_map/the_algorithms.bin
        // 0x00005555556b0000 0x00005555556bc000 0xc000             0x15a000           r--p  /home/not-matthias/Documents/work/wgit/runner/testdata/perf_map/the_algorithms.bin
        // 0x00005555556bc000 0x00005555556bf000 0x3000             0x165000           rw-p  /home/not-matthias/Documents/work/wgit/runner/testdata/perf_map/the_algorithms.bin

        const MODULE_PATH: &str = "testdata/perf_map/the_algorithms.bin";

        let (start_addr, end_addr, file_offset) = (0x00005555555a7000, 0x00005555556b0000, 0x52000);
        assert_elf_load_bias!(
            start_addr,
            end_addr,
            file_offset,
            MODULE_PATH,
            0x555555554000
        );
        insta::assert_debug_snapshot!(UnwindData::new(
            MODULE_PATH.as_bytes(),
            file_offset,
            start_addr,
            end_addr,
            None
        ));
    }

    #[test]
    fn test_ruff_unwind_data() {
        // gdb testdata/perf_map/ty_walltime -ex "break main" -ex "run" -ex "info proc mappings" -ex "continue" -ex "quit" -batch
        // 0x0000555555554000 0x0000555555e6d000 0x919000           0x0                r--p  /home/not-matthias/Documents/work/wgit/runner/testdata/perf_map/ty_walltime
        // 0x0000555555e6d000 0x0000555556813000 0x9a6000           0x918000           r-xp  /home/not-matthias/Documents/work/wgit/runner/testdata/perf_map/ty_walltime
        // 0x0000555556813000 0x00005555568a8000 0x95000            0x12bd000          r--p  /home/not-matthias/Documents/work/wgit/runner/testdata/perf_map/ty_walltime
        // 0x00005555568a8000 0x00005555568ac000 0x4000             0x1351000          rw-p  /home/not-matthias/Documents/work/wgit/runner/testdata/perf_map/ty_walltime
        // 0x00005555568ac000 0x00005555568ad000 0x1000             0x0                rw-p

        const MODULE_PATH: &str = "testdata/perf_map/ty_walltime";
        let (start_addr, end_addr, file_offset) =
            (0x0000555555e6d000_u64, 0x0000555556813000_u64, 0x918000);
        assert_elf_load_bias!(
            start_addr,
            end_addr,
            file_offset,
            MODULE_PATH,
            0x555555554000
        );

        insta::assert_debug_snapshot!(UnwindData::new(
            MODULE_PATH.as_bytes(),
            file_offset,
            start_addr,
            end_addr,
            None
        ));
    }
}
