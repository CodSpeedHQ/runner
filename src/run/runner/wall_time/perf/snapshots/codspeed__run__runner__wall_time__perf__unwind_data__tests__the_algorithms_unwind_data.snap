---
source: src/run/runner/wall_time/perf/unwind_data.rs
expression: "UnwindData::new(MODULE_PATH.as_bytes(), file_offset, start_addr, end_addr,\nNone)"
---
Ok(
    UnwindData {
        path: "testdata/perf_map/the_algorithms.bin",
        avma_range: 5555555a7000..5555556b0000,
        base_avma: 555555554000,
        eh_frame_hdr_svma: 2f0ec..33590,
        eh_frame_hdr_hash: 277fbdb59e6decaa,
        eh_frame_hash: 21d8b0c8da0d1029,
        eh_frame_svma: 128a8..2f0ec,
    },
)
